### ISSUE TITLE
Rule suggestion: no-state-in-constructor

### ISSUE THREAD

**Issue by hornta** [yannickcr/eslint-plugin-react#1810]
**INCORRECT**
```javascript
class MyComponent extends Component {
  constructor(props) {
    super(props);

    this.state = {
      counter: props.initialValue
    };
  }
}
```

**CORRECT**
```javascript
class MyComponent extends Component {
  state = {
    counter: this.props.initialValue
  };
}
```

**Why?**
It may reduce boilerplate by not having to define a constructor just for initializing the state. 

What do you think about this?

**Comment by ljharb** [comment]
Since class fields are still stage 3, and not yet supported by default eslint, I'm not sure this is worth having as a rule *yet* - but this sounds like a great rule.

**Comment by fatfisz** [comment]
Just today my workmate suggested a similar idea - forbidding `constructor` in React components altogether.

The rationale for this is that it's easy to forget to call `super` with `props`, and according to React docs [this may result in a bug](https://reactjs.org/docs/react-component.html#constructor):

>  When implementing the constructor for a React.Component subclass, you should call super(props) before any other statement. Otherwise, this.props will be undefined in the constructor, which can lead to bugs.

Then we were wondering what are the usecases for constructors, and there it is:
- Initializing properties on `this` from the arguments (which includes `state`): since React only passes one argument to the constructor, `props`, which is also available as `this.props` when initializing through class fields, we have this covered.
- Side effects: we don't want any in the constructor, so we ignore that.

So even with `constructor` we have all bases covered. As for the class fields, aren't they quite popular already in the React community? This rule wouldn't need to be turned on by default, so the decision would be up to devs. If they wanted to turn this rule on, they'd need to support class fields anyway - because otherwise where would they initialize state without a constructor?

**Comment by krawaller** [comment]
> (@fatfisz) The rationale for this is that it's easy to forget to call super with props

Note that [there is a rule](https://eslint.org/docs/rules/constructor-super) to catch omitted `super` calls.

**Comment by ljharb** [comment]
@lukyth that sounds reasonable to me! want to make the PR?

**Comment by lukyth** [comment]
Yeah. Let me try this out. I'll open a PR once I got something.
